esi_leap/api/controllers/v1/contract.py

```diff
@@ -101,14 +101,15 @@ def post(self, new_contract):
             raise exception.ContractNoOfferUUID()
 
         o_objects = offer.Offer.get(new_contract.offer_uuid_or_name)
```

Comment:
The reason Offer.get is like this is because we are using it to check how many offers are being managed by a particular project_id, but this might change from project to project, and so we first need to get all the offers by uuid or name, and do some policy checks against it. Policy checks exist in the api level so we cannot move that to the object layer

Reply:
I'm not sure I fully understand. Why does Offer.get need to check how many offers are being managed by a particular project_id?

Reply:
It doesn't, but thats what every call to Offer.get does with the results of the call.

Reply:
Okay - how does that contradict the idea that Offer.get should always only return one object, and if it matches multiple then it should throw an exception... ?

Reply:
What should it return exactly then?

Reply:
We have the get_by_name and get_by_uuid. The name function can return multiple and the uuid will only ever return one. The purpose of the get function is to get all the offers which may be identified by the given string

Reply:
Because we want to allow offer management by name or uuid, but we wanted to avoid having a uuid and a name field that can be set. So I made it one field and esi-leap will figure out if it is a name or a uuid based on the results of this Offer.get call.

Reply:
That's what I mean when I say, we can assume that `get_by_name` should always only return one result; if it returns multiple, then "matches multiple offers" Exception should be thrown. `get` returns a single item; `get_all` returns multiple. If you make this change, then all of the `o_objects` code can be reduced to `related_offer = Offer.get(...)` with some exception handling.

Reply:
but there are cases where the get_by_name should return multiple results, but the api call will still work out. Imagine two contracts created by two different projects. If both are named 'c1' the get_by_name will return multiple contracts. But because the project_id's are different for both we should expect any api calls made by the two to pass. But If an admin were to make this call, he would have management rights to both and thus would not be able to perform the api action as there are two contracts he may be talking about.

Reply:
The way to fix this would be to do policy checking inside the object layer, which we obviously do not want.

Reply:
There is no db query that can remedy this as it requires policy checks.

Reply:
`get_by_name` should never return multiple results. We can filter by `project_id` if it's non-admin or throw an exception, but that's about it.

Think of it this way: if this was a more substantial code base, and we modified `Offer.get` to potentially return multiple offers instead of just one, then *every single Offer.get call* would need to be updated as well, which is a terrible consequence. `get` was defined to return one object, while `get_all` returns multiple; it should be kept that way.

Take a look at what Ironic does:

https://opendev.org/openstack/ironic/src/branch/master/ironic/api/controllers/v1/utils.py#L212
https://opendev.org/openstack/ironic/src/branch/master/ironic/objects/node.py#L276

Note that we should also probably use that `is_uuid_like` function.

Reply:
I guess the is_uuid_like function is the way ironic deals with the ambiguity of names and uuids

Reply:
I'm not sure what .get() even should do at this point though. I think we need both get_by_name and get_by_uuid. So what is the need for get

Reply:
`get` is our function that encompasses both `get_by_uuid` and `get_by_name`; essentially what the Ironic utils function does.

Reply:
I don't know where it could ever be used though. I see how we can fix the name vs uuid part with the is_uuid_like, but I don't see how we can do the policy part.

Reply:
Are you talking about a permissions check that verifies that we're not matching a hidden offer, or an actual policy check (I don't see the latter here)? Assuming it's the former - I wonder if that's why Ironic never allowed `Node.get` to accept `uuid_or_name`, and confined that to the API controller. The solution would be to have a controller util function that checked if `uuid_or_name` is a uuid or name; if the latter, then it could check if the user is an admin, and then run `get_all` with the appropriate filters.

Reply:
Whether or not we match a certain offer is based on the name and the policy. Going back to the previous example, the admin can have access to both contrcats with the name 'c1' but the other two lessees can only access the one. This decision is made based on policy. But in order to perform this check, we need to get both of the contracts, and check their project_ids on the policy.

Reply:
get_all does not have a way to filter by a name

Reply:
Yes: I'm saying that Ironic does this in the REST API layer, and so can we. I'm saying:

a) Get rid of `get_by_uuid` and `get_by_name` and revert `get` to only accepting a uuid. `uuid_or_name` is only really relevant in the REST API; the manager service never needs to fetch anything by name.

b) Create the equivalent of Ironic's `get_rpc_node`, which does the following:

```
def get_offer_by_uuid_or_name(uuid_or_name):
   if <uuid_or_name is uuid>:
        offer = Offer.get(uuid_or_name)
   else:
         if <user is admin>:
            offers = Offer.get_all(<filter_by_name>)
         else:
             offers = Offer.get_all(<filter_by_name>, <filter_by_project_id>)
         <error handling> 
```

Reply:
c) Verify that `get_all` can filter by name. I'm pretty sure it can; the db functions pop off any filter that requires special handling, and just applies the rest generically when it calls `query = query.filter_by(**filters)`

Reply:
Would this code be in the api or the object layer.

Reply:
Well its in the api

Reply:
Fixing all of this requires a bit of refactoring. I am going to address this in the next pr after this locks one is done.

Reply:
You may want to address the refactoring first; it's better to straighten out the underlying code first.

Reply:
this has been fixed